drv/fpga-api/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+                0x0400 => Ok(FpgaError::PortDisabled),
+                0x0500 => Ok(FpgaError::NotLocked),
+                _ => Err(()),
+            },
+        }
+    }
+}
+
+impl core::convert::TryFrom<u32> for FpgaError {
+    type Error = ();
+
+    fn try_from(v: u32) -> Result<Self, Self::Error> {
+        let v: u16 = v.try_into().map_err(|_| ())?;
+        Self::try_from(v)
+    }
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum DeviceState {
+    Unknown = 0,
+    Disabled = 1,
+    AwaitingBitstream = 2,
+    RunningUserDesign = 3,
+    Error = 4,
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum BitstreamType {
+    Uncompressed = 0,
+    Compressed = 1,
+}
+
+#[derive(Copy, Clone, Debug, FromPrimitive, PartialEq, AsBytes)]
+#[repr(u8)]
+pub enum WriteOp {
+    Write = 0,
+    // This maps onto a generic Op type in Bluespec which defines Read = 1. The
+    // read/write split obviates the need for that operation.
+    BitSet = 2,
+    BitClear = 3,
+}
+
+impl From<WriteOp> for u8 {
+    fn from(op: WriteOp) -> Self {
+        op as u8
+    }
+}
+
+pub struct FpgaLock(idl::Fpga);
+
+impl Deref for FpgaLock {
+    type Target = idl::Fpga;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Drop for FpgaLock {
+    fn drop(&mut self) {
+        // We ignore the result of release because, if the server has restarted,
+        // we don't need to do anything.
+        (*self).release().ok();
+    }
+}
+
+pub struct Fpga(idl::Fpga);
+
+impl Fpga {
+    pub fn new(task_id: userlib::TaskId) -> Self {
+        Self(idl::Fpga::from(task_id))
+    }
+
+    pub fn enabled(&self) -> Result<bool, FpgaError> {
+        self.0.device_enabled()
+    }
+
+    pub fn set_enabled(&mut self, enabled: bool) -> Result<(), FpgaError> {
+        self.0.set_device_enabled(enabled)
+    }
+
+    pub fn reset(&mut self) -> Result<(), FpgaError> {
+        self.0.reset_device()
+    }
+
+    pub fn state(&self) -> Result<DeviceState, FpgaError> {
+        self.0.device_state()
+    }
+
+    pub fn id(&self) -> Result<u32, FpgaError> {
+        self.0.device_id()
+    }
+
+    pub fn start_bitstream_load(
+        &mut self,
+        bitstream_type: BitstreamType,
+    ) -> Result<Bitstream, FpgaError> {
+        let lock = self.lock()?;
+        lock.0.start_bitstream_load(bitstream_type)?;
+        Ok(Bitstream(lock))
+    }
+
+    pub fn lock(&mut self) -> Result<FpgaLock, FpgaError> {
+        self.0.lock()?;
+        Ok(FpgaLock(self.0.clone()))
+    }
+}
+
+pub struct Bitstream(FpgaLock);
+
+impl Bitstream {
+    pub fn continue_load(&mut self, data: &[u8]) -> Result<(), FpgaError> {
+        (*self.0).continue_bitstream_load(data)
+    }
+
+    pub fn finish_load(&mut self) -> Result<(), FpgaError> {
+        (*self.0).finish_bitstream_load()
+    }
+}
+
+pub struct FpgaUserDesign(idl::Fpga);
```

Comment:
Nit: It's a little confusing that this is both a `struct` in `fpga-api` and a `trait` in `fpga-driver`...